ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -0,0 +1,228 @@
+                    current <- liftIO $ readIORef inProgressVar
+                    prev <- progress style prev current t
+                    liftIO $ sleep sample
+                    progressLoop prev
+
+    progressThread <- async $ mRunLspT env $ progressLoop 0
+    let progressStop = cancel progressThread
+
+    pure ProgressReporting {..}
+
+-- | A 'ProgressReporting' that enqueues Begin and End notifications in a new
+--   thread, with a grace period (nothing will be sent if 'KickCompleted' arrives
+--   before the end of the grace period).
+--   Avoid using in tests where progress notifications are used to assert invariants.
+delayedProgressReporting
+  :: Double  -- ^ sampling rate, also used as grace period before Begin
+  -> Maybe (LSP.LanguageContextEnv c)
+  -> ProgressReportingStyle
+  -> IO ProgressReporting
+delayedProgressReporting sample lspEnv style = do
+    inProgressVar <- newVar (HMap.empty @NormalizedFilePath @Int)
+    mostRecentProgressEvent <- newTVarIO KickCompleted
+    progressAsync <- async $
+        progressThread mostRecentProgressEvent inProgressVar
+    let progressUpdate = atomically . writeTVar mostRecentProgressEvent
+        progressStop   = cancel progressAsync
+        inProgress :: NormalizedFilePath -> Action a -> Action a
+        inProgress = withProgressVar inProgressVar
+    return ProgressReporting{..}
+  where
+    -- The progress thread is a state machine with two states:
+    --   1. Idle
+    --   2. Reporting a kick event
+    -- And two transitions, modelled by 'ProgressEvent':
+    --   1. KickCompleted - transitions from Reporting into Idle
+    --   2. KickStarted - transitions from Idle into Reporting
+    progressThread mostRecentProgressEvent inProgress = progressLoopIdle
+      where
+        progressLoopIdle = do
+            atomically $ do
+                v <- readTVar mostRecentProgressEvent
+                case v of
+                    KickCompleted -> STM.retry
+                    KickStarted   -> return ()
+            asyncReporter <- async $ mRunLspT lspEnv $ do
+                -- first sleep a bit, so we only show progress messages if it's going to take
+                -- a "noticable amount of time" (we often expect a thread kill to arrive before the sleep finishes)
+                liftIO $ sleep sample
+                lspShakeProgress style inProgress
+            progressLoopReporting asyncReporter
+        progressLoopReporting asyncReporter = do
+            atomically $ do
+                v <- readTVar mostRecentProgressEvent
+                case v of
+                    KickStarted   -> STM.retry
+                    KickCompleted -> return ()
+            cancel asyncReporter
+            progressLoopIdle
+
+    lspShakeProgress style inProgress = do
+        u <- liftIO newProgressToken
+
+        void $ LSP.sendRequest LSP.SWindowWorkDoneProgressCreate
+            LSP.WorkDoneProgressCreateParams { _token = u } $ const (pure ())
```

Comment:
According to the spec, we should wait for the response to this request before proceeding. This could be done by using `signalBarrier` instead of `const (pure ())` as the response callback.